src/pybind/mgr/orchestrator.py

```diff
@@ -718,90 +763,126 @@ def validate(self, all_hosts):
-        # Maybe you're using e.g. kubenetes and you want to pass through
-        # some replicaset special sauce for autoscaling?
-        self.extended = {}
+        #: RADOS namespace where NFS client recovery data is stored in the pool.
+        self.namespace = namespace
 
-        # Object with specific settings for the service
-        self.service_spec = None
+    def validate_add(self):
+        super(NFSServiceSpec, self).validate_add()
 
-class RGWSpec(object):
+        if not self.pool:
+            raise OrchestratorValidationError('Cannot add NFS: No Pool specified')
+
+
+class RGWSpec(StatelessServiceSpec):
     """
-    Settings to configure a multisite Ceph RGW
+    Settings to configure a (multisite) Ceph RGW
     """
-    def __init__(self, hosts=None, rgw_multisite=True, rgw_zone="Default_Zone",
-              rgw_zonemaster=True, rgw_zonesecondary=False,
-              rgw_multisite_proto="http", rgw_frontend_port="8080",
-              rgw_zonegroup="Main", rgw_zone_user="zone.user",
-              rgw_realm="RGW_Realm", system_access_key=None,
-              system_secret_key=None):
-
+    # TODO: move all default values to a dedicated method. I don't want to overwrite
+    # Rook's default values.
+    def __init__(self, hosts=None, rgw_multisite=None, rgw_zone=None,
+              rgw_zonemaster=None, rgw_zonesecondary=None,
+              rgw_multisite_proto=None, rgw_frontend_port=None,
+              rgw_zonegroup=None, rgw_zone_user=None,
+              rgw_realm=None, system_access_key=None,
+              system_secret_key=None, count=None):
+
+        super(RGWSpec, self).__init__(name=rgw_zone, count=count)
+
+        #: List of hosts where RGWs should run. Not for Rook.
         self.hosts = hosts
+
         self.rgw_multisite = rgw_multisite
-        self.rgw_zone = rgw_zone
         self.rgw_zonemaster = rgw_zonemaster
         self.rgw_zonesecondary = rgw_zonesecondary
         self.rgw_multisite_proto = rgw_multisite_proto
         self.rgw_frontend_port = rgw_frontend_port
 
-        if hosts:
-            self.rgw_multisite_endpoint_addr = hosts[0]
-
-            self.rgw_multisite_endpoints_list = ",".join(
-                ["{}://{}:{}".format(self.rgw_multisite_proto,
-                                    host,
-                                    self.rgw_frontend_port) for host in hosts])
-
         self.rgw_zonegroup = rgw_zonegroup
         self.rgw_zone_user = rgw_zone_user
         self.rgw_realm = rgw_realm
 
-        if system_access_key:
-            self.system_access_key = system_access_key
-        else:
-            self.system_access_key = self.genkey(20)
-        if system_secret_key:
-            self.system_secret_key = system_secret_key
-        else:
-            self.system_secret_key = self.genkey(40)
+        self.system_access_key = system_access_key
+        self.system_secret_key = system_secret_key
+
+    def set_ansible_defaults(self):
```

Comment:
if we need to set explicitly what are the defaults depending of the orchestrator used, i think that we are to have problems.
What about to pass all this code to the __init__ method?

Reply:
Maybe you have a better idea than me. 

I'm facing the problem that Rook has it's own defaults specified within Rook. And I don't want to replace the Rook's default with new defaults defined in the orchestrator, as they might get outdated. Simply by omitting elements in the `CephObjectStore` CR, Rook will take care of setting proper defaults. 

Therefor, I'd like to know which fields are explicitly set by the user. And right now, the easiest way I could think of was to add default all parameters to `None` and only use your defaults in the ansible orchestrator.

Reply:
Then i would move the "set_ansible_defaults" method to the Ansible Orchestrator module as a function to initialize a generic RGWSpec object with the right defaults for Ansible.

Reply:
fixed